sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1891,14 +1891,34 @@ class Dataset[T] private[sql](
   * [[org.apache.spark.sql.util.QueryExecutionListener]] to the spark session.
   *
   * {{{
-  *   // Observe row count (rc) and error row count (erc) in the streaming Dataset
+  *   // Monitor the metrics for batch query using a listener.
+  *   spark.listenerManager.register(new QueryExecutionListener() {
+  *     override def onSuccess(funcName: String, qe: QueryExecution, durationNs: Long): Unit = {
+  *       qe.observedMetrics.get("my_event").foreach { row =>
```

Comment:
We should also change `observedMetrics` in `QueryExecution` to Java map to make it consistent and to support Java API properly with Java.